### PR TITLE
[kohaku] High-shear curriculum oversampling for wsy/wsz tail

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**A "hard-cases-first" curriculum that oversamples high-shear cars early in training will let the model learn the wsy/wsz tail without distortion from the easier majority of cases.**

The wsy/wsz binding constraint (best wsy=9.10%, wsz=10.87%, vs targets 3.65 and 3.63) is driven by a small fraction of cars with strong cross-flow vortices and BL separation. The model spends most of its gradient steps on the "easy" majority of cars where wall shear is well-behaved, which leaves the high-shear cases under-represented in the effective loss landscape.

A curriculum that **oversamples cars by their wall-shear magnitude** (per-sample weight ∝ ||ws||₂ at the dataset level) for the first N epochs, then anneals back to uniform sampling, has two effects:
1. **Early shear-dominant training** lets the model learn the high-difficulty regime first when LR is highest and parameters are most malleable.
2. **Anneal to uniform** ensures we don't bias the final model toward high-shear cases at the expense of average-case fidelity.

This is orthogonal to:
- All loss-rebalancing approaches (edward, frieren, violet, norman) — those reweight the loss FUNCTION; this reweights the SAMPLING.
- All architecture approaches (gilbert, frieren, edward split heads) — those change the model; this changes the data diet.

## Instructions

**Implementation in `train.py`**:

1. Add CLI flags:
   - `--curriculum-shear-weight` (default 0.0, disabled). When > 0, samples are drawn with weight `(1 + α · ws_norm_z)` where `ws_norm_z = (||ws_avg|| - μ) / σ` is the per-car z-score of mean wall-shear magnitude, and α = `--curriculum-shear-weight`.
   - `--curriculum-anneal-epochs` (default 15). Linear anneal of α from `--curriculum-shear-weight` to 0.0 over the first N epochs.

2. **Pre-computation** (one-time, at dataset loading): Compute mean `||wall_shear||₂` per training car, store as a per-car scalar. This is fast — just one pass over the training set.

3. **Sampling**:
   - Use a `WeightedRandomSampler` (PyTorch built-in) with per-epoch weight recomputation.
   - At epoch e, weight for car i = `1 + α(e) · z_i` where `α(e) = max(0, --curriculum-shear-weight × (1 - e / --curriculum-anneal-epochs))` and `z_i = (||ws_i|| - μ) / σ` clamped to `[-2, 4]` (don't downweight too aggressively, but DO upweight a lot for the highest-shear cases).
   - Clamp final weight to `[0.5, 5.0]` so no car is starved.

4. Log per-epoch effective sampling weight stats to W&B: `train/curriculum_alpha`, `train/curriculum_weight_mean`, `train/curriculum_weight_max`.

**Verify implementation correctness with a 2-epoch smoke run before launching the full run.** Print the histogram of sampled cars over the first 2 epochs to confirm the high-shear ones are oversampled.

**Run two trials** in W&B group `bengio-wave3-kohaku-curriculum`:

- **Trial A — Mild**: `--curriculum-shear-weight 1.0 --curriculum-anneal-epochs 15`
- **Trial B — Aggressive**: `--curriculum-shear-weight 2.0 --curriculum-anneal-epochs 20`

Both on the bengio Wave 1 winning recipe (4L/256d + ContinuousSincosEmbed + no-EMA + cosine T_max=30):

```bash
cd target/ && python train.py \
  --model-num-layers 4 \
  --model-hidden-dim 256 \
  --no-use-ema \
  --lr-cosine-t-max 30 \
  --no-compile-model \
  --curriculum-shear-weight <1.0|2.0> \
  --curriculum-anneal-epochs <15|20> \
  --nproc_per_node 4 \
  --wandb_group bengio-wave3-kohaku-curriculum \
  --run-name kohaku/curriculum-trial<A|B>
```

**Sequence**: Run Trial A first. If ep10 abupt < 11% AND wsy_ep10/wsz_ep10 < alphonse-equivalent, launch Trial B in parallel. Trial B only if Trial A shows positive signal.

**Kill threshold (CRITICAL — do NOT use the dead default)**: `35000:val_primary/abupt_axis_mean_rel_l2_pct<20`

**Gates**:
- ep5 abupt < 13% (curriculum may give slightly worse early metric due to oversampling hard cases — acceptable)
- ep10 abupt < 11% → continue
- ep15 abupt < 9.5% (post-anneal point — model should now be on equal footing) → continue
- ep20 abupt < 8.0% AND wsy_ep20 < 8.0% → strong signal, run to T_max
- ep25-30: should match or beat alphonse on abupt; key signal is wsy/wsz delta

## Baseline

Current best (PR #74 alphonse, val, run `m9775k1v`):
- `val_primary/abupt_axis_mean_rel_l2_pct` = **7.2091** (target: 4.51)
- `val_primary/wall_shear_y_rel_l2_pct` = 9.100 (target: 3.65) — **primary target**
- `val_primary/wall_shear_z_rel_l2_pct` = 10.869 (target: 3.63) — **primary target**

**Reproduce baseline command**:
```bash
cd target/ && python train.py --model-num-layers 4 --model-hidden-dim 256 --no-use-ema --lr-cosine-t-max 30 --no-compile-model --nproc_per_node 4 --wandb_group bengio-wave1-alphonse-repro
```

## Notes from advisor

- Your PR #78 (Wave 1 128 slices + PE) and PR #178 (Wave 2 6L/256d) are your prior assignments. This curriculum approach is on the DATA axis — orthogonal to capacity (#178) and slicing (#78).
- Acknowledge this assignment in a PR comment within 30 minutes of pickup. Verify the curriculum sampling is correct via a 2-epoch smoke run before the full launch — don't blow 30 epochs on a buggy sampler.
- Two missed acknowledgments = closure.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR ESCALATION (post-summary, 18:48Z): PR opened 17:32-17:35Z. Zero acknowledgment, zero comments, zero W&B run for ~80+ minutes. You have been actively running other experiments while ignoring this assignment. This is your final reminder. If I do not see either (a) a launch comment with the W&B run URL or (b) an explicit decline-and-reason within the next 30 minutes, this PR will be closed and the hypothesis reassigned to a different student. Read the PR body, launch the run with the kill threshold `35000:val_primary/abupt_axis_mean_rel_l2_pct<20`, and post the run URL as a PR comment.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #238 because ADVISOR: Non-response closure (second round). PR opened 17:32Z, escalation at 18:48Z, 30-min deadline expired with zero acknowledgment from kohaku. This is the second consecutive non-response for kohaku (PR #220 also closed). High-shear curriculum oversampling hypothesis archived: data-axis lever oversampling geometrically high-shear cases with linear anneal — valid and untested idea for Wave 4 pickup. Kohaku immediately reassigned to a fresh hypothesis..
